### DISCUSSION TITLE
Application OOM while resizing (mogrify) PNGs above a certain size

### DISCUSSION THREAD

**Discussion by zyqxd** [ImageMagick/ImageMagick#7461] [Help] [answered]
### ImageMagick version

7.0.8-44

### Operating system, version and so on

Oracle Linux Server 8.10

### Description

My goal is to resize the filesize of an image. I noticed PNGs take considerably more memory than JPEGs. I'd like to be able to resize these without causing my small app to fail (2gb memory). I'm running imagemagick through minimagick in a rails server

My command currently:

```
mogrify -resize 120000@
```
This ends up failing and killing my rails server (OOM)

Current resource defaults
```
bash-4.4$ /app/service/bin/identify -list resource
Resource limits:
  Width: 107.374MP
  Height: 107.374MP
  Area: 1000000000P
  List length: unlimited
  Memory: 2000000000B
  Map: 4000000000B
  Disk: 0B
  File: 786432
  Thread: 4
  Throttle: 0
  Time: 60
```

Attempt with more resources as well as setting area to 0 to force to disk. I'm okay with more time but need to process completely
```
mogrify -limit area 0 -limit disk 8GiB -limit map 256MiB -limit memory 128MiB -resize 120000@ /app/tmp/mini_magick20240704-2708461-uj0s4o.png

failed with status: 1 and error:
mogrify: CacheResourcesExhausted `/app/tmp/mini_magick20240704-2708461-uj0s4o.png' @ error/cache.c/OpenPixelCache/3860.
```

I think theres something about the PNG format that I may not know, so looking for some guidance on that. FWIW this is the image that is killing my server:

![original](https://github.com/minimagick/minimagick/assets/3641047/38c98867-8b20-456e-ae4d-a918d71b65f9)

**Comment by fmw42** [comment]
> mogrify -resize 120000@

This is not a valid command. It has no input image.

For a single image, please use

`magick input.suffix -resize 120000@ output.suffix`

or 

`magick mogrify -resize 120000@ image.suffix`

however, magick mogrify is generally not to be used for a single image. It was designed to be used on a folder of images.  Also magick is the more flexible command is true IM 7, where magick mogrify is mogrify from IM 6 run via IM 7.

**Comment by snibgo** [comment, marked as answer]
>Resource limits:
  Disk: 0B

That limit is probably set in your policy.xml. A command `-limit` option can tighten a limit, but can't relax it.

> mogrify -limit area 0 -limit disk 8GiB

`-limit disk 8GiB` can't relax a limit of zero bytes.

`-limit area` limits the size in pixels (width * height) of images that can be processed in memory. Limiting this to zero pixels means any pixel cache will be written to disk. But your disk limit is zero. Hence you can't do anything.

**Reply by zyqxd** [reply]
Is this the case for ENV variables as well? If I set for my session MAGICK_DISK_LIMIT=128GiB, would this relax the disk limit? Or is policy.xml the only place where this can be increased?

Also, do you have documentation on this? There are a lot of threads on stack overflow that mentions using limit in the case of increasing memory which seem misleading if this is the case

**Reply by zyqxd** [reply]
Ah, I found it 

> Note, you can restrict limits relative to any [security policies](https://imagemagick.org/source/policy-open.xml), but you cannot relax them.


https://imagemagick.org/script/command-line-options.php#limit

**Comment by snibgo** [comment]
Limits can be set system-wide (in `policy.xml`, typically in the same directory as `magick`), and specifically for a user (in the user's `policy.xml`), and in environment variables, and in `magick` commands and other IM commands.

These methods are symmetrical: the most severe of all these limits are the ones used, and no method can relax a limit set by another method. So if the most severe of them specify (for example) a disk limit of 1 GiB, then no IM process will use more than 1 GiB of disk for pixel caches.

The command `magick -list resource` will say what the current limits are, but don't say what has caused the limit.
